### PR TITLE
Fix S4143 FP with Indexers

### LINKED ISSUE #4178: Fix S4143 FP: Indexers
### Description

S4143 False positive when using indexers.

### Repro steps

```cs
var first = new Curve();
var second = new Curve();

first.Points[0] = new Point();
second.Points[0] = new Point();
```
S4143 is reported, although the index in accessed on separate objects.

### Expected behavior

No warning reported.

### Actual behavior

S4143 is reported.

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #4178

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.cs [hunk 1]
```diff
@@ -300,4 +300,22 @@ public void Foo()
...
+            var first = new ClassWithListProperty();
+            var second = new ClassWithListProperty();
+
+            first.IntList[0] = 1;
+            second.IntList[0] = 2;
+        }
+
+        private class ClassWithListProperty
+        {
+            public List<int> IntList { get; set; }
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.cs:318]: could you also add IntList2, and then the test above should be:

```
            first.IntList[0] = 1;
            first.IntList2[0] = 11;
            second.IntList[0] = 2;
            second.IntList2[0] = 22;
```
and no issue should be raised

Also, could you also add separate tests with various combinations of correctly raised issues like

```
            first.IntList[0] = 1;
            first.IntList2[0] = 2;
            first.IntList[0] = 3; // Noncompliant
            first.IntList2[1] = 4; // ok, different index
```

and 

```
            first.IntList[0] = 1;
            first.IntList2[1] = 2;
            first.IntList[1] = 3; 
            first.IntList2[1] = 4; // Noncompliant
```

and more compliant test cases like:
```
            first.IntList[0] = 1;
            first.IntList2[1] = 2;
            first.IntList[1] = 3; 
            first.IntList2[0] = 4;
```

Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.cs:318]: also, can you add some tests with nested objects, like

```
private class ClassWithNested
{
            public ClassWithListProperty Alpha { get; set; }
            public ClassWithListProperty Beta { get; set; }
}
```

and something like

```
            first.Alpha.IntList[0] = 1;
            first.Alpha.IntList2[0] = 11;
            first.Beta.IntList[0] = 1111;
            first.Beta.IntList2[0] = 11111;
            second.Alpha.IntList[0] = 2;
             second.Alpha.IntList2[0] = 22;
            second.Beta.IntList[0] = 222;
            second.Beta.IntList2[0] = 2222;
```

and a noncompliant case with such nested structure

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotOverwriteCollectionElements.cs [hunk 2]
```diff
@@ -155,9 +152,9 @@ private static SyntaxNode GetIdentifier(ExpressionSyntax expression)
                     // ElementAccess    // <-- we are here
                     //   Arguments: index
                     //   Expression: SimpleMemberAccess
-                    //                 Expression: a
-                    //                 Name: b  // <-- we need this
-                    return ((MemberAccessExpressionSyntax)expression).Name;
+                    //                 Expression: a // <-- we need this
+                    //                 Name: b
+                    return ((MemberAccessExpressionSyntax)expression);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotOverwriteCollectionElements.cs:157]: We don't need to change the type. And there's extra pair of parenthesis.
```
return expression;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotOverwriteCollectionElements.cs [hunk 3]
```diff
@@ -155,9 +152,9 @@ private static SyntaxNode GetIdentifier(ExpressionSyntax expression)
                     // ElementAccess    // <-- we are here
                     //   Arguments: index
                     //   Expression: SimpleMemberAccess
-                    //                 Expression: a
-                    //                 Name: b  // <-- we need this
-                    return ((MemberAccessExpressionSyntax)expression).Name;
+                    //                 Expression: a // <-- we need this
+                    //                 Name: b
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotOverwriteCollectionElements.cs:156]: This comment is not correct. We don't need `a`. We need `a.b` and that matches the implementation.

**Pull request merged**